### DISCUSSION TITLE
Why does the top_k value significantly affect the retrieve results in RAG ?

### DISCUSSION THREAD

**Discussion by qiulang** [open-webui/open-webui#6210] [Q&A]
During my test I find the top_k value significantly affect the retrieve results in RAG. 

When I set top_k to 8, I got the correct result. The page 3 with the score 0.9955145716667175 and the page 1 with the score 0.9836521148681641 are the two correct contexts to my question.

```
  metadatas: [
    [
      {
        file_id: '8d96a203-5a60-4389-bedc-ba28097c72be',
        page: 0,
        start_index: 0,
        score: 0.9962517619132996,
      },
      {
        file_id: '8d96a203-5a60-4389-bedc-ba28097c72be',
        page: 3,
        start_index: 0,
        score: 0.9955145716667175,
      },
      {
        file_id: '8d96a203-5a60-4389-bedc-ba28097c72be',
        page: 5,
        start_index: 0,
        score: 0.9903665781021118,
      },
      {
        file_id: '8d96a203-5a60-4389-bedc-ba28097c72be',
        page: 1,
        start_index: 0,
        score: 0.9836521148681641,
      },
...
```

Then I set the top_k to 6 and asked the same question. But this time the page 3 (the most relevant document chunk)does not show up in the top 6 candidates.

```
metadatas: [
    [
      {
        file_id: '8d96a203-5a60-4389-bedc-ba28097c72be',
        page: 0,
        start_index: 0,
        score: 0.9962517619132996,
      },
      {
        file_id: '8d96a203-5a60-4389-bedc-ba28097c72be',
        page: 5,
        start_index: 0,
        score: 0.9903665781021118,
      },
      {
        file_id: '8d96a203-5a60-4389-bedc-ba28097c72be',
        page: 1,
        start_index: 0,
        score: 0.9836521148681641,
      },
...
```

Why is that ?

**Comment by thiswillbeyourgithub** [comment]
I don't think it should change the result either.

Are you using a reranker ? Have you tried disabling it? Are you asking the exact same question ?

**Reply by qiulang** [reply]
Turn off rerank does not make the difference (yes I ask the exact same question, otherwise I would not raise this question).

BTW, the result I attached was using rerank, turning off rerank, the result would show in different format. 

I attach the screenshot below in case other may have the same doubt

**Comment by qiulang** [comment]
<img width="1115" alt="top 8" src="https://github.com/user-attachments/assets/f00d1801-b6ad-4484-b6c4-72702a5896a0">

VS

<img width="1113" alt="top 6" src="https://github.com/user-attachments/assets/414aa7cc-d1a7-4609-b197-47d84756f627">

**Comment by qiulang** [comment]
With more documents adding to the knowledge base, I need to increase the top k value to get the retrieve results right again.

For example, I added my documents to 20+ documents (not a big number at all) and the top 8 results fail to retrieve the relevant results but if I changed to top 10 I can get the relevant results right.

**Reply by thiswillbeyourgithub** [reply]
This indicates that the embedding model is not doing a good job. Either because the model is bad or because its bad on your language specifically or there's a bug in the code that makes bad document used as input to the model. That's my guess at least. Have you tried with openai's paid embeddings to compare?

**Reply by peuportier** [reply]
To enhance document retrieval capabilities, I integrated the Apache Tika engine as a supplementary tool. This will allow for more efficient extraction of metadata and content from various document formats. Additionally, to improve search results, I will increase the top_k parameter to 10, which will return the top 10 most relevant documents for each query.
For documents containing Chinese characters, I recommend utilizing a Chinese embedding model, such as BERT or Word2Vec, specifically trained on Chinese text data. This will enable the retrieval system to better understand the nuances of the Chinese language and provide more accurate search results.
The combination of the Tika engine and a Chinese embedding model will significantly enhance the document retrieval capabilities, particularly for large collections of documents. By leveraging these technologies, users can expect more precise and relevant search results, even when dealing with complex queries or large volumes of data."

**Reply by qiulang** [reply]
For Chinese language, of course we should use Chinese embedding. I use `BAAI/bge-large-zh-v1.5` for embedding and `BAAI/bge-reranker-base` for rerank.  My question is Why does the top_k value significantly affect the retrieve results. I may try Tika engine  later but my question remains.

**Reply by peuportier** [reply]
When top_k is set to a higher value, more results are retrieved, which increases the chances of finding relevant documents but also includes more potentially less relevant results. Conversely, a smaller top_k value will return fewer results, which may lead to missing out on relevant documents if they are ranked lower. **" the top_k parameter controls how many of the most similar items are returned from the database or index of embeddings. "**

In your case, using BAAI/bge-reranker-base means that after retrieving the top_k documents, the reranker adjusts their ranking based on a more sophisticated model. A larger top_k gives the reranker more results to choose from, potentially improving the quality of the final ranked list.

Hope it's help you to understand

**Reply by qiulang** [reply]
Thanks for message but I have to say you did not get my question. You may check @thiswillbeyourgithub 's comment, he got it (even though he probably does not speak Chinese.)

**Reply by qiulang** [reply]
Besides, I think if we can have 2 top_K values, one for initial retrieval result the other for re-rank result we probably can get a better result. But currently open-webui does not support that.

**Reply by thiswillbeyourgithub** [reply]
@peuportier to me the good question raised by @qiulang is that the actual score shown in the logs should not change if we change top_k

I agree with the idea to use 2 top_k. Actually for the llm the order matters not much so using a reranker to just resort the results is heresy.

And yep, can't read chinese at all

**Reply by qiulang** [reply]
Right and as my comment to you said, to further debug the issue I turned off rerank, but different top_K values still affected the retrieval scores. That really looks like a bug, either in webui or langchain.

**Comment by qiulang** [comment]
@thiswillbeyourgithub @peuportier I am embarrassed  to say @peuportier was actually right and @thiswillbeyourgithub was probably mislead by my words. The reason I dismissed peuportier's words too quickly was because you tried to explain to me what rerank and top_k mean and of course I knew what they mean so I just jumped to the conclusion that you didn't get my question. 

But after rethinking about it and did more test (I have tested them almost 100 hundred times and I am really sick of those questions now) I figured out the reason. It also shows some defect in the current RAG implementation in my opinion. 

Let me break it down, first the reason to my problem,

1. when I import about 10 articles I need to set top_k to 8 to make my RAG work, setting top_k to 6 fails to answer my question. The reason is because the chunk data that contains my answer in at top 7 at retrieval stage. After rerank it became **top 2**, so top 6 fails to answer my question.
2. when I import about 20 articles then ask the same question, that chunk data is now at top 9 at retrieval stage, so setting top_k to 8 fails to work again.

The defect I see are,

1. Only do I turn on hybrid search can I use rerank. But I would like to use rerank even without hybrid. I check the code and see hybrid uses BM25, basically a keyword search as,

```
        ensemble_retriever = EnsembleRetriever(
            retrievers=[bm25_retriever, chroma_retriever], weights=[0.5, 0.5]
        )
```

But there is no reason why I can't use rerank without hybrid.

2. I hope to see 2 top_k, one for retrieval and the other for rerank.  In my case, at retrieval stage the distance for each candidate is close, so I would like a bigger top k values for retrieval. Then a smaller top k value for rerank. In my case top 10 for retrieval and top 5 for rerank is good combination.

**Comment by oatmealm** [comment]
In case this helps, I've noticed recently the reranker (any I've used) return no documents when the text splitting is not character (default).

**Reply by qiulang** [reply]
For what it's worth, I have long given up open-webui for my RAG app, and find other project works for me.

**Reply by mahenning** [reply]
I can not confirm this. I use hybrid search w/ BAAI/bge-reranker-v2-m3 reranker and Tiktoken splitter and get RAG results with knowledge base and also with drop-in-chat documents.

**Reply by iratipj** [reply]
@mahenning Which embedding are you using?

**Reply by mahenning** [reply]
`intfloat/multilingual-e5-base` with 512 chunk size and 64 token overlap. The e5 model truncates the input after 512 token.

**Reply by choicebai86** [reply]
> For what it's worth, I have long given up open-webui for my RAG app, and find other project works for me.

Would you be so gracious as to divulge the other projects that might be more fitting? I have had the pleasure of employing RAGFLOW, and it has imparted a rather commendable performance.

**Reply by qiulang** [reply]
@choicebai86 I did try ragflow, but ragflow does not have logs to show what kinds questions/answers my client has asked (or any log at all), so if you want to use it as a commercial product it may be a show stopper. I have raised a feature to ragflow and they just said they would consider. So I also tried others.
